exonum-java-binding-core/rust/integration_tests/benches/jni_cache.rs

```diff
@@ -20,22 +20,98 @@ extern crate java_bindings;
 extern crate lazy_static;
 extern crate test;
 
-use integration_tests::mock::transaction::create_mock_transaction_proxy;
-use integration_tests::vm::create_vm_for_benchmarks_with_fakes;
-use java_bindings::exonum::blockchain::Transaction;
-use java_bindings::jni::JavaVM;
-use java_bindings::MainExecutor;
+use integration_tests::{
+    mock::transaction::create_mock_transaction_proxy, vm::create_vm_for_benchmarks_with_fakes,
+};
+use java_bindings::{
+    exonum::blockchain::Transaction,
+    jni::{objects::JObject, JNIEnv, JavaVM},
+    utils::{convert_to_string, get_class_name, jni_cache},
+    JniExecutor, JniResult, MainExecutor,
+};
 
 use std::sync::Arc;
 use test::{black_box, Bencher};
 
+const ARITHMETIC_EXCEPTION_CLASS: &str = "java/lang/ArithmeticException";
+
 lazy_static! {
     pub static ref VM: Arc<JavaVM> = create_vm_for_benchmarks_with_fakes();
-    pub static ref EXECUTOR: MainExecutor = MainExecutor::new(VM.clone());
+    pub static ref EXECUTOR: MainExecutor = {
+        let ex = MainExecutor::new(VM.clone());
+        ex.with_attached(|env| {
+            jni_cache::init_cache(env);
+            Ok(())
+        })
+        .unwrap();
+        ex
+    };
+}
+
+// Returns a class name of an obj as a `String`. It's a simulation of old non cached implementation.
+fn get_class_name_not_cached_impl(env: &JNIEnv, obj: JObject) -> JniResult<String> {
+    let class_object = env
+        .call_method(obj, "getClass", "()Ljava/lang/Class;", &[])?
+        .l()?;
+    let class_name = env
+        .call_method(class_object, "getName", "()Ljava/lang/String;", &[])?
+        .l()?;
+    convert_to_string(env, class_name)
+}
+
+fn create_exception<'a>(env: &'a JNIEnv) -> JObject<'a> {
+    env.new_object(ARITHMETIC_EXCEPTION_CLASS, "()V", &[])
+        .unwrap()
+}
+
+#[bench]
+pub fn is_instance_of_cached(b: &mut Bencher) {
+    EXECUTOR
+        .with_attached(|env| {
+            let exception = create_exception(env);
+            let tx_exec_exception_ref = jni_cache::classes_refs::transaction_execution_exception();
+
+            b.iter(|| black_box(env.is_instance_of(exception, &tx_exec_exception_ref)));
+            Ok(())
+        })
+        .unwrap();
+}
+
+#[bench]
+pub fn is_instance_of_not_cached(b: &mut Bencher) {
+    EXECUTOR
+        .with_attached(|env| {
+            let exception = create_exception(env);
+            b.iter(|| black_box(env.is_instance_of(exception, ARITHMETIC_EXCEPTION_CLASS)));
```

Comment:
Does `iter` perform measurements immediately (i.e., when you call `iter`, not when this code returns)?

Also, as this code does not check for results, it may return `Err`s and one will never notice, will they?

Reply:
The `.iter()` method measures time spent to iterations. And that summary data gets processed (average, min, max) after return from this method.
And the method doesn't check for the result. But in our case, there should always be `true` I believe.
The general recommendation for writing benchmarks using the standard Rust abilities is to put all setup code outside of the `.iter()` method and only put the part one wants to measure inside.

Reply:
> And the method doesn't check for the result.

Shan't it `unwrap` to ensure that we do get _some_ results and not an error, which is somewhat easy with JNI?